src/tools/rbd/action/Ubbd.cc

```diff
@@ -0,0 +1,190 @@
+  std::string pool_name;
+  std::string nspace_name;
+  std::string image_name;
+  std::string snap_name;
+  struct ubbdd_mgmt_rsp rsp = { 0 };
+  struct ubbd_map_options opts = { 0 };
+  std::string dev_type = "rbd";
+
+  int r = utils::get_pool_image_snapshot_names(
+    vm, at::ARGUMENT_MODIFIER_NONE, &arg_index, &pool_name, &nspace_name,
+    &image_name, &snap_name, true, utils::SNAPSHOT_PRESENCE_PERMITTED,
+    utils::SPEC_VALIDATION_NONE);
+  if (r < 0) {
+    return r;
+  }
+  opts.type = opts.generic_dev.opts.type = dev_type.c_str();
+  opts.generic_dev.opts.rbd.pool = pool_name.c_str();
+  opts.generic_dev.opts.rbd.ns = nspace_name.c_str();
+  opts.generic_dev.opts.rbd.image = image_name.c_str();
+  opts.generic_dev.opts.rbd.snap = snap_name.c_str();
+
+  if (vm["read-only"].as<bool>()) {
+    opts.read_only = true;
+  }
+  if (vm["exclusive"].as<bool>()) {
+	  opts.generic_dev.opts.rbd.exclusive = true;
+  }
+  if (vm["quiesce"].as<bool>()) {
+	  opts.generic_dev.opts.rbd.quiesce = true;
+  }
+  if (vm.count("quiesce-hook")) {
+	  opts.generic_dev.opts.rbd.quiesce_hook = vm["quiesce-hook"].as<std::string>().c_str();
+  }
+
+  r = ubbd_map(&opts, &rsp);
+  if (r < 0) {
+    std::cerr << "rbd: ubbd device map failed: " << cpp_strerror(r) << std::endl;
+    return r;
+  }
+
+  std::cout << rsp.add.path << std::endl;
+
+  return 0;
+#else
+  std::cerr << "rbd: ubbd device is not supported" << std::endl;
+  return -EOPNOTSUPP;
+#endif
+}
+
+#if defined(WITH_RBD_UBBD)
+static int parse_unmap_options(const std::string &options_string,
+		struct ubbd_unmap_options *unmap_opts)
+{
+  char *options = strdup(options_string.c_str());
+  BOOST_SCOPE_EXIT(options) {
+    free(options);
+  } BOOST_SCOPE_EXIT_END;
+
+  for (char *this_char = strtok(options, ", ");
+       this_char != NULL;
+       this_char = strtok(NULL, ",")) {
+    char *value_char;
+
+    if ((value_char = strchr(this_char, '=')) != NULL)
+      *value_char++ = '\0';
+
+    if (!strcmp(this_char, "force")) {
+      unmap_opts->force = true;
+    } else {
+      std::cerr << "rbd: ubbd unknown unmap option '" << this_char << "'"
+                << std::endl;
+      return -EINVAL;
+    }
+  }
+
+  return 0;
+}
+
+int execute_unmap(const po::variables_map &vm,
+                  const std::vector<std::string> &ceph_global_init_args) {
+  std::string device_name = utils::get_positional_argument(vm, 0);
+  struct ubbdd_mgmt_rsp rsp = { 0 };
+  struct ubbd_unmap_options opts = { 0 };
+  int r;
+
+  if (!boost::starts_with(device_name, "/dev/ubbd")) {
+    std::cerr << "rbd: unmapping a ubbd device needs a device path (/dev/ubbdX)" << std::endl;
+    return -EINVAL;
+  }
+
+  device_name.erase(0, 9);
+  opts.ubbdid = stoi(device_name);
+
+  if (vm.count("options")) {
+    for (auto &options : vm["options"].as<std::vector<std::string>>()) {
+      r = parse_unmap_options(options, &opts);
+      if (r < 0) {
+        std::cerr << "rbd: ubbd couldn't parse unmap options" << std::endl;
```

Comment:
nit: this error message looks redundant. I think we may just rely on the `parse_unmap_options` will print a meaningful message in the case of error.

Reply:
ok, sounds good